cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/ProgramRunDispatcherTask.java

```diff
@@ -0,0 +1,152 @@
+import io.cdap.cdap.api.plugin.Requirements;
+import io.cdap.cdap.api.service.worker.RunnableTask;
+import io.cdap.cdap.api.service.worker.RunnableTaskContext;
+import io.cdap.cdap.app.guice.AppFabricServiceRuntimeModule;
+import io.cdap.cdap.app.runtime.Arguments;
+import io.cdap.cdap.app.runtime.ProgramController;
+import io.cdap.cdap.app.runtime.ProgramOptions;
+import io.cdap.cdap.app.runtime.ProgramRunnerFactory;
+import io.cdap.cdap.common.app.RunIds;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.internal.app.ApplicationSpecificationAdapter;
+import io.cdap.cdap.internal.app.deploy.ConfiguratorFactory;
+import io.cdap.cdap.internal.app.deploy.InMemoryProgramRunDispatcher;
+import io.cdap.cdap.internal.app.deploy.RemoteProgramRunDispatcherResponse;
+import io.cdap.cdap.internal.app.deploy.pipeline.ProgramRunInfo;
+import io.cdap.cdap.internal.app.runtime.artifact.ApplicationClassCodec;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactRepository;
+import io.cdap.cdap.internal.app.runtime.artifact.RequirementsCodec;
+import io.cdap.cdap.internal.app.runtime.codec.ArgumentsCodec;
+import io.cdap.cdap.internal.app.runtime.codec.ProgramOptionsCodec;
+import io.cdap.cdap.internal.io.SchemaTypeAdapter;
+import io.cdap.cdap.security.impersonation.Impersonator;
+import org.apache.twill.api.RunId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+
+/**
+ * Implementation of {@link RunnableTask} to execute Program-run operation in a system worker.
+ */
+public class ProgramRunDispatcherTask implements RunnableTask {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramRunDispatcherTask.class);
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(
+          new GsonBuilder())
+      .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+      .registerTypeAdapter(ApplicationClass.class, new ApplicationClassCodec())
+      .registerTypeAdapter(Requirements.class, new RequirementsCodec())
+      .registerTypeAdapter(RunId.class, new RunIds.RunIdCodec())
+      .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
+      .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec())
+      .create();
+
+  private final Injector injector;
+
+  @Inject
+  ProgramRunDispatcherTask(Injector injector) {
+    this.injector = injector;
+  }
+
+  @Override
+  public void run(RunnableTaskContext context) throws Exception {
+    try {
+      ProgramRunInfo programRunInfo = GSON.fromJson(context.getParam(), ProgramRunInfo.class);
+      DispatchTaskRunner taskRunner = injector.getInstance(DispatchTaskRunner.class);
+      ProgramController programController = taskRunner.dispatch(programRunInfo);
+      RemoteProgramRunDispatcherResponse response = null;
+      if (Objects.nonNull(programController)) {
```

Comment:
when will `programController ` be null?

Reply:
Rearranged the code to make it more readable. ProgramController might be null if we are unable to launch a pipeline.